src/cockpit/channels/http.py

```diff
@@ -0,0 +1,153 @@
+                else:
+                    context = ssl.create_default_context(cafile=opt_tls['authority']['file'])
+            else:
+                context = ssl.create_default_context()
+
+            if 'validate' in opt_tls and not opt_tls['validate']:
+                context.check_hostname = False
+                context.verify_mode = ssl.VerifyMode.CERT_NONE
+
+            connection = http.client.HTTPSConnection(opt_address, opt_port, context=context)
+        else:
+            connection = http.client.HTTPConnection(opt_address, opt_port)
+
+        try:
+            if opt_unix:
+                # create the connection's socket so that it won't call .connect() internally (which only supports TCP)
+                connection.sock = socket.socket(socket.AF_UNIX)
+                connection.sock.connect(opt_unix)
+            else:
+                # explicitly call connect(), so that we can do proper error handling
+                connection.connect()
+        except (OSError, IOError) as e:
+            logger.error('Failed to open %s:%s: %s %s', opt_address, opt_port or opt_unix, type(e), e)
+            problem = 'unknown-hostkey' if isinstance(e, ssl.SSLCertVerificationError) else 'not-found'
+            self.close(problem=problem, message=str(e))
+            return None
+
+        return connection
+
+    def read_send_response(self, response):
+        '''Completely read the response and send it to the channel'''
+
+        while True:
+            # we want to stream data blocks as soon as they come in
+            block = response.read1(4096)
+            if not block:
+                logger.debug('reading response done')
+                # this returns immediately and does not read anything more, but updates the http.client's
+                # internal state machine to "response done"
+                block = response.read()
+                assert block == b''
+                break
+            logger.debug('read block of size %i', len(block))
+            self.loop.call_soon_threadsafe(self.send_data, block)
+
+    def parse_headers(self, http_msg):
+        headers = dict(http_msg)
+        remove = ['Connection', 'Transfer-Encoding']
+        if self.options.get('binary'):
+            remove = ['Content-Length', 'Range']
+        for h in remove:
+            try:
+                del headers[h]
+            except KeyError:
+                pass
+        return headers
+
+    def request(self):
+        connection = self.create_connection()
+        if not connection:
+            # make_connection does the error reporting
+            return
+
+        connection.request(self.options.get('method'),
+                           self.options.get('path'),
+                           headers=self.options.get('headers') or {},
+                           body=self.body)
+        response = connection.getresponse()
+        self.loop.call_soon_threadsafe(lambda: self.send_control(
```

Comment:
Are `.call_soon_threadsafe()` calls guaranteed to be well-ordered when they arrive?

Reply:
The [call_soon()](https://docs.python.org/3/library/asyncio-eventloop.html?highlight=call%20soon%20threadsafe#asyncio.loop.call_soon) docs say "Callbacks are called in the order in which they are registered", and c_s_t() says "A thread-safe variant of call_soon()", which I interpret as "the "output order is the same as the input order". If that weren't so, this API would be useless at least for our cases indeed. But it at least survived the "highly parallel torture test" as well as c-podman.